data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1256 @@
+    priorities: VecDeque<Priorities>,
+}
+
+impl PriorityEngine {
+    /// Retrieve the inner fee information as a vector.
+    pub fn as_vec(&self) -> Vec<Priorities> {
+        self.priorities.iter().rev().cloned().collect_vec()
+    }
+
+    /// Provide suggestions for sensible transaction priority values, together with their expected
+    /// time-to-block in epochs.
+    ///
+    /// This is only a first approach to an estimation algorithm. There is abundant prior art about
+    /// fee estimation in other blockchains. We might revisit this once we collect more insights
+    /// about our fees market and user feedback.
+    pub fn estimate_priority(
+        &self,
+        seconds_per_epoch: Duration,
+    ) -> Result<PrioritiesEstimate, PriorityError> {
+        // Short-circuit if there are too few tracked epochs for an accurate estimation.
+        let len = u32::try_from(self.priorities.len())?;
+        if len < MINIMUM_TRACKED_EPOCHS {
+            Err(PriorityError::NotEnoughSampledEpochs(
+                len,
+                MINIMUM_TRACKED_EPOCHS,
+                (MINIMUM_TRACKED_EPOCHS - len) * u32::try_from(seconds_per_epoch.as_secs())? / 60
+                    + 1,
+            ))?;
+        }
+
+        Ok(strategies::target_minutes(
+            self.priorities.iter(),
+            [360, 60, 15, 5, 1],
+            u16::try_from(seconds_per_epoch.as_secs())?,
+        ))
+    }
+
+    /// Creates a new engine with the default capacity from a vector of `Priorities`.
+    ///
+    /// This assumes that the vector is ordered from newest to oldest.
+    pub fn from_vec(priorities: Vec<Priorities>) -> Self {
+        Self::from_vec_with_capacity(priorities, DEFAULT_QUEUE_CAPACITY_EPOCHS)
+    }
+
+    /// Creates a new engine with a custom queue capacity from a vector of `Priorities`.
+    ///
+    /// This assumes that the vector is ordered from newest to oldest.
+    pub fn from_vec_with_capacity(priorities: Vec<Priorities>, capacity: usize) -> Self {
+        // Create a new queue with the desired capacity
+        let mut fees = VecDeque::with_capacity(capacity);
+        // Push as many elements from the input as they can fit in the queue
+        priorities.into_iter().take(capacity).for_each(|entry| {
+            fees.push_back(entry);
+        });
+
+        Self { priorities: fees }
+    }
+
+    /// Get the entry at a certain position, if an item at that position exists, or None otherwise.
+    #[cfg(test)]
+    #[inline]
+    pub fn get(&self, index: usize) -> Option<&Priorities> {
+        self.priorities.get(index)
+    }
+
+    /// Push a new `Priorities` entry into the engine.
+    #[inline]
+    pub fn push_priorities(&mut self, priorities: Priorities) {
+        log::trace!("Pushing new transaction priorities entry: {:?}", priorities);
+        // If we hit the capacity limit, pop from the back first so the queue does not grow
+        if self.priorities.len() + 1 == self.priorities.capacity() {
```

Comment:
This is wrong, if the capacity is 2047 we want to store at most 2047 items, right? With the +1 here it will never store more than 2046, because 2046 + 1 == 2047 so it will pop one element making the final length 2046 again.

```suggestion
        if self.priorities.len() == self.priorities.capacity() {
```

Also there is no guarantee that `self.priorities.capacity()` will be the same capacity as the one passed to `VecDeque::with_capacity(capacity)`, because it is allowed to be bigger. To ensure a fixed sized queue we need to store an additional `capacity` parameter in the `PriorityEngine` struct.

Reply:
> This is wrong, if the capacity is 2047 we want to store at most 2047 items, right? With the +1 here it will never store more than 2046, because 2046 + 1 == 2047 so it will pop one element making the final length 2046 again.

That's what I believed, but test `engine_capacity` proved the contrary!

> Also there is no guarantee that `self.priorities.capacity()` will be the same capacity as the one passed to `VecDeque::with_capacity(capacity)`, because it is allowed to be bigger. To ensure a fixed sized queue we need to store an additional `capacity` parameter in the `PriorityEngine` struct.

Please elaborate, I don't get it :sweat_smile:

Reply:
> That's what I believed, but test engine_capacity proved the contrary!

That test works by pure chance, if you make a similar test with a capacity to 3 it won't work.

> Please elaborate, I don't get it

```rust
let buf: VecDeque<i32> = VecDeque::with_capacity(10);
assert_eq!(buf.capacity(), 10);
```

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `15`,
 right: `10`', src/main.rs:4:1
```